prov/efa/test/efa_unit_test_rdm_peer.c

```diff
@@ -0,0 +1,135 @@
+	/* Packet type must be >= EFA_RDM_REQ_PKT_BEGIN */
+	efa_unit_test_eager_msgrtm_pkt_construct(pkt_entry, &pkt_attr);
+
+	(&peer->robuf)->exp_msg_id = exp_msg_id;
+	ret = efa_rdm_peer_reorder_msg(peer, efa_rdm_ep, pkt_entry);
+	assert_int_equal(ret, expected_ret);
+	(&peer->robuf)->exp_msg_id = 0;
+
+	if (msg_id >= efa_env.recvwin_size + exp_msg_id) {
+		/* Check the overflow_pke_list contains the overflow msg_id */
+		assert_int_equal(efa_unit_test_get_dlist_length(&peer->overflow_pke_list), 1);
+		dlist_foreach_container_safe (
+			&peer->overflow_pke_list,
+			struct efa_rdm_peer_overflow_pke_list_entry,
+			overflow_pke_list_entry, entry, tmp) {
+			overflow_pkt_entry = overflow_pke_list_entry->pkt_entry;
+			overflow_msg_id = ((struct efa_rdm_rtm_base_hdr *) overflow_pkt_entry->wiredata)->msg_id;
+			assert_int_equal(overflow_msg_id, msg_id);
+			/* Clean up */
+			dlist_remove(&overflow_pke_list_entry->entry);
+			efa_rdm_pke_release_rx(overflow_pke_list_entry->pkt_entry);
+			ofi_buf_free(overflow_pke_list_entry);
+		}
+	} else {
+		efa_rdm_pke_release_rx(pkt_entry);
+	}
+}
+
+void test_efa_rdm_peer_reorder_expected_msg_id(struct efa_resource **state) {
+	struct efa_resource *resource = *state;
+	u_int32_t msg_id, exp_msg_id;
+	int expected_ret;
+
+	efa_unit_test_resource_construct(resource, FI_EP_RDM);
+
+	msg_id = 0;
+	exp_msg_id = 0;
+	expected_ret = 0;
+	/* Receiving expected message id should return 0 */
+	test_efa_rdm_peer_reorder_msg_impl(resource, exp_msg_id, msg_id, expected_ret);
+}
+
+
+void test_efa_rdm_peer_reorder_smaller_msg_id(struct efa_resource **state) {
+	struct efa_resource *resource = *state;
+	u_int32_t msg_id, exp_msg_id;
+	int expected_ret;
+
+	efa_unit_test_resource_construct(resource, FI_EP_RDM);
+
+	msg_id = 1;
+	exp_msg_id = 10;
+	expected_ret = -FI_EALREADY;
+	/* Receiving message id smaller than expected should return -FI_EALREADY */
+	test_efa_rdm_peer_reorder_msg_impl(resource, exp_msg_id, msg_id, expected_ret);
+}
+
+void test_efa_rdm_peer_reorder_larger_msg_id(struct efa_resource **state) {
+	struct efa_resource *resource = *state;
+	u_int32_t msg_id, exp_msg_id;
+	int expected_ret;
+
+	efa_unit_test_resource_construct(resource, FI_EP_RDM);
+
+	msg_id = 10;
+	exp_msg_id = 0;
+	expected_ret = 1;
+	efa_env.rx_copy_ooo = 0; /* Do not copy this pkt entry */
+	/* Receiving message id larger than expected should return 1 */
+	test_efa_rdm_peer_reorder_msg_impl(resource, exp_msg_id, msg_id, expected_ret);
+}
+
+void test_efa_rdm_peer_reorder_overflow_msg_id(struct efa_resource **state) {
+	struct efa_resource *resource = *state;
+	u_int32_t msg_id, exp_msg_id;
+	int expected_ret;
+
+	efa_unit_test_resource_construct(resource, FI_EP_RDM);
+
+	msg_id = 16384;
+	exp_msg_id = 0;
+	expected_ret = 1;
+	/* Message id that overflows the receive window should be put in the
+	 * overflow_pke_list and return 1 */
+	test_efa_rdm_peer_reorder_msg_impl(resource, exp_msg_id, msg_id, expected_ret);
+}
```

Comment:
Can we have more unit tests to cover

1.  moving entry from overflow to recv window
2. two pkt entries of same msg id will be appended to the same entry.